scalameta/quasiquotes/shared/src/main/scala-3/scala/meta/internal/quasiquotes/ReificationMacros.scala

```diff
@@ -0,0 +1,688 @@
+      case '{ ($sc: StringContext).self } => (sc, QuasiquoteType.Self)
+      case '{ ($sc: StringContext).template } => (sc, QuasiquoteType.Template)
+      case '{ ($sc: StringContext).mod } => (sc, QuasiquoteType.Mod)
+      case '{ ($sc: StringContext).enumerator } => (sc, QuasiquoteType.Enumerator)
+      case '{ ($sc: StringContext).importer } => (sc, QuasiquoteType.Importer)
+      case '{ ($sc: StringContext).importee } => (sc, QuasiquoteType.Importee)
+      case '{ ($sc: StringContext).source } => (sc, QuasiquoteType.Source)
+      case _ => quotes.reflect.report.errorAndAbort("Expected call to quasiquote extension method. ")
+    new ReificationMacros().expandUnapply(stringContext, scrutineeExpr, quasiquoteType)
+}
+
+class ReificationMacros(using val topLevelQuotes: Quotes) {
+  rei =>
+  import scala.meta.Dialect
+  import scala.meta.inputs.{Position => MetaPosition, _}
+  import scala.meta.{Tree => MetaTree}
+
+  import topLevelQuotes.reflect._
+  type MetaParser = (Input, Dialect) => MetaTree
+  lazy val RootPackage = topLevelQuotes.reflect.Symbol.classSymbol("_root_")
+
+  private sealed trait Mode {
+    def isTerm: Boolean = this.isInstanceOf[Mode.Term]
+    def isPattern: Boolean = this.isInstanceOf[Mode.Pattern]
+    def hasHoles: Boolean = this match
+      case Mode.Term(_, holes) => !holes.isEmpty
+      case Mode.Pattern(_, holes, _) => !holes.isEmpty
+
+    def multiline: Boolean
+  }
+  private object Mode {
+    case class Term(multiline: Boolean, holes: List[TermHole]) extends Mode
+    case class Pattern(multiline: Boolean, holes: List[PatternHole], unapplySelector: Expr[Any])
+        extends Mode
+  }
+  private case class TermHole(name: String, arg: Term, var reifier: Option[Term])
+  private case class PatternHole(
+      name: String,
+      posStart: Int,
+      posEnd: Int,
+      tpe: Option[TypeRepr],
+      var symbol: Option[Symbol],
+      var reifier: Option[Term]
+  )
+
+  lazy val ListClass = TypeRepr.of[scala.List]
+
+  def expandUnapply(
+      strCtx: Expr[StringContext],
+      unapplySelector: Expr[Any],
+      qType: QuasiquoteType
+  ): Expr[Any] = {
+    val mode = extractModePattern(strCtx, unapplySelector)
+    expand(strCtx, qType, mode)
+  }
+
+  def expandApply(
+      strCtx: Expr[StringContext],
+      args: Expr[Seq[Any]],
+      qType: QuasiquoteType
+  ): Expr[Any] = {
+    val mode = extractModeTerm(strCtx, args)
+    expand(strCtx, qType, mode)
+  }
+
+  private def expand(strCtx: Expr[StringContext], qType: QuasiquoteType, mode: Mode) = {
+    val input = metaInput()
+    val (dialect, dialectExpr) = instantiateDialect(mode)
+    val parser = instantiateParser(qType)
+    val skeleton = parseSkeleton(parser, input, dialect)
+    reifySkeleton(skeleton, mode, qType, dialectExpr, input)
+  }
+
+  private def metaInput() = {
+    val pos = Position.ofMacroExpansion
+    val reflectInput = pos.sourceFile
+    val content = new String(reflectInput.content.get)
```

Comment:
why do you need `new String`? isn't `.content` an `Option[String]` already?

also, what to do if it's None?